pkg/pillar/evetpm/tpm.go

```diff
@@ -389,8 +385,225 @@ func readDiskKey(log *base.LogObject) ([]byte, error) {
 	keyBytes, err := tpm2.NVReadEx(rw, TpmDiskKeyHdl,
 		tpm2.HandleOwner, EmptyPassword, 0)
 	if err != nil {
-		log.Errorf("NVReadEx failed: %v", err)
-		return nil, err
+		return nil, fmt.Errorf("NVReadEx failed: %v", err)
 	}
 	return keyBytes, nil
 }
+
+//FetchSealedVaultKey fetches Vault key sealed into TPM2.0
+func FetchSealedVaultKey() ([]byte, error) {
+
+	//gain some knowledge about existing environment
+	sealedKeyPresent := isSealedKeyPresent()
+	legacyKeyPresent := isLegacyKeyPresent()
+
+	if !sealedKeyPresent && !legacyKeyPresent {
+		//Fresh install, go for a new key
+		key, err := GetRandom(vaultKeyLength)
+		if err != nil {
+			return nil, fmt.Errorf("FetchSealedVaultKey: GetRandom failed, %v", err)
+		}
+		err = SealDiskKey(key, DiskKeySealingPCRs)
+		if err != nil {
+			return nil, fmt.Errorf("FetchSealedVaultKey: Sealing failed: %v", err)
+		}
+	}
+
+	if !sealedKeyPresent && legacyKeyPresent {
+		//XXX: we need a migration path for existing installations.
+		//hence re-using the current key here. i.e. if we end up creating
+		//a new random key here, and we fail the upgrade, the fallback
+		//image will not be able to unlock the vault, and we will not be
+		//able to do any upgrade either (since the base image is downloaded
+		//to vault as well).
+		//Hence, we do it in two-steps:
+		//First clone the existing key(not sealed), and make it a sealed key.
+		//In a subsequent release, rotate the sealed key to a new one.
+		//Upgrade path will be to first upgrade to a) first release and then b)
+		key, err := readDiskKey()
+		if err != nil {
+			return nil, fmt.Errorf("Error in retrieving old key")
+		}
+		err = SealDiskKey(key, DiskKeySealingPCRs)
+		if err != nil {
+			return nil, fmt.Errorf("FetchSealedVaultKey: Sealing failed: %v", err)
+		}
+	}
+	//sealedKeyPresent && !legacyKeyPresent : unseal
+	//sealedKeyPresent && legacyKeyPresent  : unseal
+
+	//By this, we have a key sealed into TPM
+	return UnsealDiskKey(DiskKeySealingPCRs)
+}
+
+//SealDiskKey seals key into TPM2.0, with provided PCRs
+func SealDiskKey(key []byte, pcrSel tpm2.PCRSelection) error {
+	rw, err := tpm2.OpenTPM(TpmDevicePath)
+	if err != nil {
+		return err
+	}
+	defer rw.Close()
+
+	tpm2.NVUndefineSpace(rw, EmptyPassword,
+		tpm2.HandleOwner, TpmSealedDiskPubHdl)
+
+	tpm2.NVUndefineSpace(rw, EmptyPassword,
+		tpm2.HandleOwner, TpmSealedDiskPrivHdl)
+
+	session, policy, err := PolicyPCRSession(rw, pcrSel)
+	if err != nil {
+		return fmt.Errorf("PolicyPCRSession failed: %v", err)
+	}
+
+	//Don't need the handle, we need only the policy for sealing
+	if err := tpm2.FlushContext(rw, session); err != nil {
+		return fmt.Errorf("Unable to flush session handle %v: %v", session, err)
+	}
+
```

Comment:
If the device power cycles e.g. here, will we be able to recover the information from the TPM? We got rid of the passwords  above AFAIU

Reply:
Not a problem, since the key supplied here can be 
1) FetchSealedDiskKey(), which will again supply either same value (if cloning it from non-sealed copy) or generate a new random value again (which is again ok, since we are yet to setup the vault in that case)
or 
2) From Controller post attestation, in which case, we will again get same key back post-reboot as well

Reply:
Added this note as a comment in code